### ISSUE TITLE
Details on "Reimplement the C# API" for 0.15

### ISSUE THREAD

**Issue by christophwille** [bytecodealliance/wasmtime-dotnet#9]
I was just updating the NuGet from 0.12 and literally everything turned red. To clarify the API changes I'll take this as an example:

https://github.com/christophwille/csharp-opa-wasm/blob/master/src/Opa.Wasm/OpaHost.cs

No more IHost, no more Import - basically everything is now runtime on Host?

https://github.com/christophwille/csharp-opa-wasm/blob/master/src/Opa.Wasm.ConsoleSample/Program.cs

No more Engine, no more Store? 

Given that the API changed in such a big way, can you describe a bit about why now this way compared to before? (I'd like to fix my code in a way that matches your API). Thanks.

**Comment by peterhuene** [comment]
Sure, and sorry about the big change, but it should be pretty straight forward to port from 0.12.0.

Reflection (and thus "IHost") is no longer used to link host-provided functions, globals, and memories.

Instead, you create an instance of `Host` which has functions for defining functions, globals, and memories, as well as a WASI instance.

Take a quick look at the [memory example](https://github.com/bytecodealliance/wasmtime-dotnet/blob/master/examples/memory/Program.cs) which showcases most of the changes.

The `DefineFunction` method takes any `Action` or `Func` that has WebAssembly types (i.e. int/long/float/double), as well as a special `Caller` type that must be the first parameter if used.

The `Caller` class has methods to get an exported memory from the caller so you can get data being passed to it.

Let me know if you have problems updating your source, I'd be happy to help with anything you run into.  Sorry again about the big change, but hopefully this makes the API a little easier to use with increased functionality (specifically, the linking details are now all in Wasmtime itself, so you can easily link in different WASI implementations from the engine as well as provide your own functions).

**Comment by peterhuene** [comment]
Also, the [API docs](https://bytecodealliance.github.io/wasmtime-dotnet/api/index.html) are now current with 0.15.0.

**Comment by christophwille** [comment]
I'll go through my codebase later today and let you know if I came across any stumbling blocks.

**Comment by christophwille** [comment]
So porting is easy https://github.com/christophwille/csharp-opa-wasm/commit/3f0eac99860b8e44330044946881dfbfbf19761d

One thing I noticed in the benchmark though is that it is slower now. Before, I was loading engine, store, module - and it created the host for the test. Now, the host loads the module, which is like in total 30 times slower, given the old perf https://github.com/christophwille/csharp-opa-wasm/tree/e74ea738214c3d944c6311c5c84dab1186d92290/spikes/BenchmarkOpaWasm this is quite an increase.

What could I properly cache in concurrent scenarios?

**Comment by peterhuene** [comment]
So currently hosts and their instances have thread-affinity, meaning they're not safe to interact with from other threads (this is a documented requirement of the C API and we haven't fully fleshed out the Rust API's Sync/Send on the various engine primitives just yet.

The change to hide the store under the `Host` should not impact performance, but it does change the use pattern which I think is the source of the problem.

For your use case, I think you should create a `Host` only once in `WasmPolicyExecution` which defines the memory and host functions that the modules will use.  You should also only load the target module into the host a single time.

Then, in the benchmark function, you would start by zeroing the host's memory (`DefineMemory` returns a reference to it you can use to get the span to zero it out).  Then you would call `Host.Instantiate` to run the sample application.  This would closer approximate what the 0.12.0 code was doing; it might even be slightly faster as we only have to set the `Host` up once to run multiple instances.

Does that make sense?

**Comment by christophwille** [comment]
I'll rephrase to see if I got this correct (for simplicity assuming a single wasm):

* the perf I am seeing is for end-to-end newing a Host, defining memory, defining functions, instantiating an instance and running ("one off").
* a shortcut would be to reuse an existing Host *after* a previous run, zeroing the Memory, and the instantiating a new instance ("reuse")
* also, I could create a pool of similarly configured Host's to dish out available Hosts in a concurrent scenario.
* one wasm only (because others will have differing imports et al).

You might have seen my ASP.NET AuthZ spike in my repo (and ultimately what got me started thinking about OPA in .NET). This will have many different wasms (compiled regos) for different authz policy scenarios - what would be a good scenario then if can't reuse most anything?

The current ~40ms make it more palatable to call out to Gatekeeper.

**Comment by peterhuene** [comment]
> the perf I am seeing is for end-to-end newing a Host, defining memory, defining functions, instantiating an instance and running ("one off").

That's correct.  Previously you created a single engine and store, where a single module was loaded into.  Most of the perf cost comes from loading the module since that is when the module gets Just-In-Time compiled.  With the changes, the host hides the engine and store implementation since most use cases were 1:1 engine and store.  As a result, loading the module is done in the inner loop resulting in the noticable slow down.

> a shortcut would be to reuse an existing Host after a previous run, zeroing the Memory, and then instantiating a new instance ("reuse")

That's correct.  If the host is reused, the module is compiled only once.  Instantiating the module is a very cheap operation by comparison.

> also, I could create a pool of similarly configured Host's to dish out available Hosts in a concurrent scenario.

I believe that could work.

> one wasm only (because others will have differing imports et al).

If you desire different imports, a different host would be required or you can call [`ClearDefinitions`](https://bytecodealliance.github.io/wasmtime-dotnet/api/Wasmtime.Host.html#Wasmtime_Host_ClearDefinitions) to reset what the host currently defines.

Unfortunately I'm not at all familiar with AuthZ or OPA, so I can't say what would make the most sense for you.  However, I feel like this indicates the .NET API isn't quite there yet to cover the case where you want to provide per-instantiation differentiation.

Let me think more on this and see how we can better adapt the API for this use case.

**Comment by peterhuene** [comment]
At the very least, I think perhaps exposing the store (again) where you can load modules into independent of the host would make sense.

**Comment by christophwille** [comment]
Being able to inject the Store into a different Host (or something similar) sounds like it would get the most perf given your previous explanations. Unless it isn't thread safe of course.

In my OPA case, getting access to the Memory from the instance would do the trick (in my current implementation, I can only get access to the Memory in BuildHost() and have to keep a reference, because the Instance gives me only a MemoryImport)

**Comment by christophwille** [comment]
@peterhuene Do you have plans to share/discuss how (if at all) you intend to change the API?

**Comment by peterhuene** [comment]
Hi @christophwille.  Sorry for the slow response.  I haven't spent time on designing this yet, but I'm going to work on this now.   I'll get back to you soon regarding any potential changes.

**Comment by peterhuene** [comment]
I propose the following *breaking* changes:

Split the `Host` abstraction into two abstractions: `Store` (this is where you load modules) and `Host` (this is where you instantiate modules).  Hosts would accept a store upon construction.

This would allow for you to create a single store to load one or more modules, but then instantiate the same module into different hosts.  Obviously the hosts would need to define the same set of expected imports for the module, but the implementations could then differ.

**Comment by christophwille** [comment]
As long as the most expensive part (module compilation) happens only once I am happy. If the Store doesn't do that - then there won't be any real savings, right?

**Comment by peterhuene** [comment]
@christophwille sorry for the delay.  I've put up #21 that should help.

The `Store` type is back and is where modules get loaded.  A `Host` is constructed from a `Store`:

From the `hello` example:

```c#
using var store = new Store();
using var module = store.LoadModuleText("hello.wat");
using var host = new Host(store);

host.DefineFunction(
    "",
    "hello",
    () => Console.WriteLine("Hello from C#, WebAssembly!")
);

using dynamic instance = host.Instantiate(module);
...
```

Thus, a module can be loaded once into a store, but any number of hosts may make use of that module.

Does this address your issue?

**Comment by christophwille** [comment]
I have a couple questions that might be useful for others too:

1. The compile hit is incurred where? On LoadModule*, on new Host() or Instantiate(...)? 
2. store.LoadModule* - what is the behavior when I load the same module again? Concurrently even? (read: do I have to lock loading reentrancy which obviously could happen in the Web API case?)
3. Unrelated... any plans to ever let me define the entire API surface beforehand? Goal: not using dynamic.

**Comment by peterhuene** [comment]
1. `LoadModule` compiles the module.  `Instantiate` is a much faster operation that just allocates what is needed to run an instance of the module.  Allocating a `Host` internally creates a new *linker*, which is only used for resolving imports to instantiated modules.
1. Stores and their associated objects should only be accessed by a single thread.  The Wasmtime C API has a way of sharing a module between threads, but it isn't exposed in the .NET API yet. 
1.  Once the *interface types* proposal is fully implemented in Wasmtime, I would love to see a way to generate the API surface from an interface definition.  That said, the `DynamicObject` support is just a convenience over finding the function and invoking it.  If you plan on repeatedly calling an exported function, I would find it in `Instance.Externs`, cache it, and call `Invoke` on it directly.

**Comment by christophwille** [comment]
Sounds good. That should get rid of my main perf bottleneck that was in 0.15.

Having #2 would be great, #3 really only matters in super high-perf scenarios (but I can envision those)

**Comment by peterhuene** [comment]
@christophwille I've pushed 0.18.0 of the package with the corresponding Wasmtime version and these changes.  Let me know if you run into any problems.  Thanks!

**Comment by christophwille** [comment]
I took the 0.18 package - please see https://github.com/christophwille/csharp-opa-wasm/issues/8 if you happen to have an idea what sort of problem I stumbled upon this time around... the code is only minimally changed from 0.15 that only performed slowly but didn't crash.

**Comment by peterhuene** [comment]
I've put up a fix at https://github.com/bytecodealliance/wasmtime-dotnet/pull/24.  When that gets merged, I'll release 0.18.1 for you.
